src/esphome/climatedevice/climatedevice.h

```diff
@@ -0,0 +1,232 @@
+#ifndef ESPHOME_CLIMATEDEVICE_CLIMATEDEVICE_H
+#define ESPHOME_CLIMATEDEVICE_CLIMATEDEVICE_H
+
+#include "esphome/defines.h"
+
+#ifdef USE_CLIMATEDEVICE
+
+#include "esphome/component.h"
+#include "esphome/automation.h"
+#include "esphome/esppreferences.h"
+#include "esphome/climatedevice/climatedevice_traits.h"
+
+ESPHOME_NAMESPACE_BEGIN
+
+namespace climatedevice {
+
+/// Simple enum to represent the mode of a climate device.
+enum ClimateDeviceMode {
+  CLIMATEDEVICE_MODE_OFF = 0,   ///< The climate device is off.
+  CLIMATEDEVICE_MODE_AUTO = 1,  ///< The climate device is in auto mode.
+  CLIMATEDEVICE_MODE_COOL = 2,  ///< The climate device is in cool mode.
+  CLIMATEDEVICE_MODE_HEAT = 3,  ///< The climate device is in heat mode.
+};
+
+template<typename... Ts> class ControlAction;
+template<typename... Ts> class PublishAction;
+template<typename... Ts> class CurrentTemperatureAction;
+
+struct ClimateDeviceState {
```

Comment:
This struct should bundle the different state variables of the climate device. It is a state parameter for the control lambda function and is created by the StateCall class to provide an atomic state change.